File: src/test/java/org/algohub/engine/compiler/java/MemoryJavaCompilerTest.java
package org.algohub.engine.compiler.java;

import org.junit.Test;

import java.lang.reflect.Method;

import static org.assertj.core.api.Assertions.assertThat;

public class MemoryJavaCompilerTest {

    private static final String SOURCE_CODE =
            "public final class Solution {\n" +
                    "public static String greeting(String name) {\n" +
                    "\treturn \"Hello \" + name;\n" + "}\n}\n";

    @Test
    public void compileAndRunStaticMethod() throws Throwable {

        Object[] greetingObject = MemoryJavaCompiler.INSTANCE.compileMethod(
                "Solution", "greeting", SOURCE_CODE
        );

        Object obj = greetingObject[0];
        Method greeting = (Method) greetingObject[1];

        Object result = greeting.invoke(obj, "Julia");
        assertThat(result).isEqualTo("Hello Julia");
    }
}

File: src/test/kotlin/com/jalgoarena/judge/MemoryJavaCompilerTest.kt
package com.jalgoarena.judge

import com.jalgoarena.compile.CompileErrorException
import com.jalgoarena.compile.MemoryJavaCompiler
import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class MemoryJavaCompilerTest {

    private val SOURCE_CODE =
            "public final class Solution {\npublic static String greeting(String name) {\n\treturn \"Hello \" + name;\n}\n}\n"

    @Test
    fun compileAndRunStaticMethod() {

        val (instance, method) = MemoryJavaCompiler().compileMethod(
                "Solution", "greeting", SOURCE_CODE
        )

        val result = method.invoke(instance, "Julia")
        assertThat(result).isEqualTo("Hello Julia")
    }

    @Test(expected = NoSuchMethodError::class)
    fun throwsNoSuchMethodExceptionWhenInvokingNonExistingMethod() {
        MemoryJavaCompiler().compileMethod(
                "Solution", "dummy", SOURCE_CODE
        )
    }

    @Test(expected = IllegalStateException::class)
    fun throwsIllegalStateExceptionWhenTryingToInvokeClassWithPrivateConstructor() {
        val sourceCode = "public final class Solution { private Solution() { }; public void dummy() {} }"

        MemoryJavaCompiler().compileMethod(
                "Solution", "dummy", sourceCode
        )
    }

    @Test(expected = CompileErrorException::class)
    fun throwsCompileErrorExceptionWhenSourceCodeDoesNotCompile() {
        val sourceCodeMissingReturnString = "public final class Solution { private Solution() { }; public String dummy() {} }"

        MemoryJavaCompiler().compileMethod(
                "Solution", "dummy", sourceCodeMissingReturnString
        )
    }
}
